util-core/src/main/scala/com/twitter/concurrent/AsyncStream.scala

```diff
@@ -121,12 +100,16 @@ final class AsyncStream[+A] private (private val underlying: Future[LazySeq[A]])
    *   AsyncStream(1, 2, 3).takeWhile(_ < 0) = AsyncStream.empty
    * }}}
    */
-  def takeWhile(p: A => Boolean): AsyncStream[A] = AsyncStream(
-    foldRight(empty[A].underlying) { (a, as) =>
-      if (p(a)) Future.value(Cons(a, AsyncStream(as)))
-      else empty.underlying
+  def takeWhile(p: A => Boolean): AsyncStream[A] =
+    this match {
+      case Empty => empty
+      case FromFuture(fa) => M(fa.map(a => if (p(a)) this else empty))
```

Comment:
You mean like M(fa.map { a => if (p(a)) this else empty })? Oh I see, it could help differentiate the M parens and the map closure...